addons/hr_payroll/models/hr_payslip.py

```diff
@@ -0,0 +1,570 @@
+            return False
+        return True
+
+    @api.multi
+    def cancel_sheet(self):
+        return self.write({'state': 'cancel'})
+
+    @api.multi
+    def process_sheet(self):
+        return self.write({'paid': True, 'state': 'done'})
+
+    @api.multi
+    def hr_verify_sheet(self):
+        self.compute_sheet()
+        return self.write({'state': 'verify'})
+
+    @api.multi
+    def refund_sheet(self):
+        for payslip in self:
+            id_copy = payslip.copy({'credit_note': True, 'name': _('Refund: %s') % payslip.name})
+            payslip.signal_workflow('hr_verify_sheet')
+            payslip.signal_workflow('process_sheet')
+
+        form_res = self.env.ref('hr_payroll.view_hr_payslip_form').id or False
+        tree_res = self.env.ref('hr_payroll.view_hr_payslip_tree').id or False
+        return {
+            'name':_("Refund Payslip"),
+            'view_mode': 'tree, form',
+            'view_id': False,
+            'view_type': 'form',
+            'res_model': 'hr.payslip',
+            'type': 'ir.actions.act_window',
+            'target': 'current',
+            'domain': "[('id', 'in', %s)]" % [id_copy.id],
+            'views': [(tree_res, 'tree'), (form_res, 'form')],
+            'context': {}
+        }
+
+    @api.multi
+    def unlink(self):
+        for payslip in self.filtered(lambda x: x.state not in ['draft','cancel']):
+            raise UserError(_('You cannot delete a payslip which is not draft or cancelled'))
+        return super(HrPayslip, self).unlink()
+
+    #TODO move this function into hr_contract module, on hr.employee object
+    def get_contract(self, employee, date_from, date_to):
+        """
+        @param employee: browse record of employee
+        @param date_from: date field
+        @param date_to: date field
+        @return: returns the ids of all the contracts for the given employee that need to be considered for the given dates
+        """
+        #a contract is valid if it ends between the given dates
+        clause_1 = ['&',('date_end', '<=', date_to),('date_end','>=', date_from)]
+        #OR if it starts between the given dates
+        clause_2 = ['&',('date_start', '<=', date_to),('date_start','>=', date_from)]
+        #OR if it starts before the date_from and finish after the date_end (or never finish)
+        clause_3 = ['&',('date_start','<=', date_from),'|',('date_end', '=', False),('date_end','>=', date_to)]
+        clause_final =  [('employee_id', '=', employee.id),'|','|'] + clause_1 + clause_2 + clause_3
+        return self.env['hr.contract'].search(clause_final)
+
+    @api.multi
+    def compute_sheet(self):
+        HrPayslipLine = self.env['hr.payslip.line']
+        IrSequence = self.env['ir.sequence']
+        for payslip in self:
+            number = payslip.number or IrSequence.next_by_code('salary.slip')
+            #delete old payslip lines
+            old_sliplines = HrPayslipLine.search([('slip_id', '=', payslip.id)])
```

Comment:
cannot we directly use one2many reference instead of search :  like
payslip.line_ids